### PR TITLE
apple update shutdown option.

### PR DESCRIPTION
**Pull request by weswhet**
adding in the option to have munki shutdown after apple updates instead of restarting.

Should fix an issue with T2 machines needing to be shut down instead of doing a normal restart.

### CODE REVIEW

**Comment by weswhet** [comment]
Just getting the code on the internet. About to do some various tests.

File: code/client/munkilib/authrestart/__init__.py [hunk 1]
```diff
@@ -181,13 +181,20 @@ def perform_auth_restart(username=None, password=None):
...
+def do_authorized_or_normal_restart(username=None,
+                                    password=None,
+                                    shutdown=False):
+    '''Do a shutdown if needed, or an authrestart if allowed/possible,
+    else do a normal restart.'''
     display.display_info('Restarting now.')
+    if shutdown:
+        # we need a shutdown here instead of any type of restart
+        display.display_debug1('Performing a shutdown...')
+        dummy_retcode = subprocess.call(['/sbin/shutdown', '-h', '-o', 'now'])
```
Comment (gregneagle) [inline code/client/munkilib/authrestart/__init__.py]: Why don't we just return after this and get rid of all the changes below here?

File: code/client/munkilib/authrestart/__init__.py [hunk 2]
```diff
@@ -181,13 +181,20 @@ def perform_auth_restart(username=None, password=None):
...
 
 
-def do_authorized_or_normal_restart(username=None, password=None):
-    '''Do an authrestart if allowed/possible, else do a normal restart.'''
+def do_authorized_or_normal_restart(username=None,
+                                    password=None,
+                                    shutdown=False):
+    '''Do a shutdown if needed, or an authrestart if allowed/possible,
+    else do a normal restart.'''
     display.display_info('Restarting now.')
```
Comment (gregneagle) [inline code/client/munkilib/authrestart/__init__.py]: Seems like this should be moved to after the `if shutdown` block

**Comment by weswhet** [comment]
Below logs are from a successful test. 

```
Dec 07 2018 13:30:05 -0800 ### Starting managedsoftwareupdate run: logoutinstall ###
Dec 07 2018 13:30:05 -0800 Starting...
Dec 07 2018 13:30:05 -0800     The following Apple Software Updates are available to install:
Dec 07 2018 13:30:05 -0800         + macOS 10.14.2 Update-
Dec 07 2018 13:30:05 -0800            *Restart required
Dec 07 2018 13:30:05 -0800 Installing available Apple Software Updates...
Dec 07 2018 13:30:05 -0800     Caching CatalogURL https://swscan.apple.com/content/catalogs/others/index-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog
Dec 07 2018 13:30:06 -0800 DEBUG2: Options: {'logging_function': <function display_debug2 at 0x103201b18>, 'ignore_system_proxy': False, 'additional_headers': {'User-Agent': u'managedsoftwareupdate/3.6.0.3663 Darwin/18.2.0 (x86_64) (MacBookAir8,1)'}, 'file': '/tmp/munki_swupd_cache/mirror/apple.sucatalog.download', 'cache_data': {
    etag = "\"5aed87-57c718231be40\"";
    "last-modified" = "Fri, 07 Dec 2018 16:57:21 GMT";
}, 'url': 'https://swscan.apple.com/content/catalogs/others/index-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog', 'follow_redirects': True, 'download_only_if_changed': True, 'can_resume': True}
Dec 07 2018 13:30:06 -0800 DEBUG2: URLSession_task_didReceiveChallenge_completionHandler_
Dec 07 2018 13:30:06 -0800 DEBUG2: Authentication challenge for Host: swscan.apple.com Realm: None AuthMethod: NSURLAuthenticationMethodServerTrust
Dec 07 2018 13:30:06 -0800 DEBUG2: Allowing OS to handle authentication request
Dec 07 2018 13:30:07 -0800 DEBUG1: Status: 304
Dec 07 2018 13:30:07 -0800 DEBUG1: Headers: {u'Expires': u'Fri, 07 Dec 2018 21:31:27 GMT', u'Last-Modified': u'Fri, 07 Dec 2018 16:57:21 GMT', u'Connection': u'keep-alive', u'Etag': u'"5aed87-57c718231be40"', u'Cache-Control': u'max-age=81', u'Date': u'Fri, 07 Dec 2018 21:30:06 GMT', u'Content-Type': u'application/x-apple-plist'}
Dec 07 2018 13:30:07 -0800 DEBUG1: Item is unchanged on the server.
Dec 07 2018 13:30:07 -0800 DEBUG1: /tmp/munki_swupd_cache/mirror/apple.sucatalog already exists and is up-to-date.
Dec 07 2018 13:30:07 -0800 DEBUG1: softwareupdate cmd: ['/usr/local/munki/ptyexec', '/usr/sbin/softwareupdate', '--verbose', '-i', u'macOS 10.14.2 Update- ']
Dec 07 2018 13:30:09 -0800 Installing macOS 10.14.2 Update
Dec 07 2018 13:30:38 -0800     Done.
Dec 07 2018 13:30:39 -0800     To install these updates, your computer must shut down. Your computer will automatically start up to finish installation.
Dec 07 2018 13:30:39 -0800     Installation will not complete successfully if you choose to restart your computer instead of shutting down.
Dec 07 2018 13:30:39 -0800     Please call halt(8) or select Shut Down from the Apple menu. To automate the shutdown process with softwareupdate(8), use --restart.
Dec 07 2018 13:30:39 -0800 DEBUG2: This update requires a shutdown...
Dec 07 2018 13:30:39 -0800 DEBUG2: InstallHistory.plist items:
[{
    date = "2018-12-07 21:30:08 +0000";
    displayName = "macOS 10.14.2 Update";
    displayVersion = " ";
    packageIdentifiers =     (
        "com.apple.pkg.macOSBrain"
    );
    processName = softwareupdated;
}]
Dec 07 2018 13:30:39 -0800 DEBUG1: Raw Apple Update install results: {'download': [], 'failures': [], 'installed': []}
Dec 07 2018 13:30:39 -0800 DEBUG2: Matched macOS 10.14.2 Update in InstallHistory.plist
Dec 07 2018 13:30:39 -0800 DEBUG2: Found shutdown flag...
Dec 07 2018 13:30:39 -0800 Finishing...
Dec 07 2018 13:30:39 -0800 Saving application inventory...
Dec 07 2018 13:30:39 -0800 DEBUG1: Getting info on currently installed applications...
Dec 07 2018 13:30:48 -0800 ### Ending managedsoftwareupdate run ###
Dec 07 2018 13:30:48 -0800 Software installed or removed requires a restart.
Dec 07 2018 13:30:53 -0800 DEBUG1: Performing a shutdown...
```

File: code/client/munkilib/authrestart/__init__.py [hunk 3]
```diff
@@ -190,7 +190,7 @@ def do_authorized_or_normal_restart(username=None,
         # we need a shutdown here instead of any type of restart
         display.display_debug1('Performing a shutdown...')
         dummy_retcode = subprocess.call(['/sbin/shutdown', '-h', '-o', 'now'])
-        return dummy_retcode
+        return ''
```
Comment (gregneagle) [inline code/client/munkilib/authrestart/__init__.py]: Nope. Just `return`. No return value at all.

**Comment by weswhet** [comment]
Successfully updated 2 more machines using this build pkg. 
[munkitools-3.6.0.3675.pkg.zip](https://github.com/munki/munki/files/2665130/munkitools-3.6.0.3675.pkg.zip)

**Comment by erikng** [comment]
Alright I tested on a 2018 T2 15" MBP. Working as intended though I have some comments.

```
Dec 10 2018 16:57:14 -0600 ### Starting managedsoftwareupdate run: logoutinstall ###
Dec 10 2018 16:57:14 -0600 Starting...
Dec 10 2018 16:57:14 -0600     Performing preflight tasks...
Dec 10 2018 16:57:17 -0600 DEBUG2: curl: (28) Operation timed out after 2006 milliseconds with 0 out of 0 bytes received
Dec 10 2018 16:57:17 -0600 DEBUG2: 
Dec 10 2018 16:57:17 -0600 DEBUG2: Didn't receive valid JSON.
Dec 10 2018 16:57:17 -0600     preflight stdout: 'cache' had errors during execution!
Dec 10 2018 16:57:17 -0600     The following Apple Software Updates are available to install:
Dec 10 2018 16:57:17 -0600         + macOS 10.14.2 Update-
Dec 10 2018 16:57:17 -0600            *Restart required
Dec 10 2018 16:57:17 -0600 Installing available Apple Software Updates...
Dec 10 2018 16:57:17 -0600     Caching CatalogURL https://swscan.apple.com/content/catalogs/others/index-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog
Dec 10 2018 16:57:18 -0600 DEBUG2: Options: {'logging_function': <function display_debug2 at 0x1079f9b18>, 'ignore_system_proxy': False, 'additional_headers': {'User-Agent': u'managedsoftwareupdate/3.6.0.3675 Darwin/18.2.0 (x86_64) (MacBookPro15,1)'}, 'file': '/tmp/munki_swupd_cache/mirror/apple.sucatalog.download', 'cache_data': {
Dec 10 2018 16:57:18 -0600 DEBUG2: URLSession_task_didReceiveChallenge_completionHandler_
Dec 10 2018 16:57:18 -0600 DEBUG2: Authentication challenge for Host: swscan.apple.com Realm: None AuthMethod: NSURLAuthenticationMethodServerTrust
Dec 10 2018 16:57:18 -0600 DEBUG2: Allowing OS to handle authentication request
Dec 10 2018 16:57:18 -0600 DEBUG1: Status: 304
Dec 10 2018 16:57:18 -0600 DEBUG1: Headers: {u'Expires': u'Mon, 10 Dec 2018 23:05:56 GMT', u'Last-Modified': u'Mon, 10 Dec 2018 20:51:18 GMT', u'Connection': u'keep-alive', u'Etag': u'"5aed87-57cb120637d80"', u'Cache-Control': u'max-age=518', u'Date': u'Mon, 10 Dec 2018 22:57:18 GMT', u'Content-Type': u'application/x-apple-plist'}
Dec 10 2018 16:57:18 -0600 DEBUG1: Item is unchanged on the server.
Dec 10 2018 16:57:18 -0600 DEBUG1: /tmp/munki_swupd_cache/mirror/apple.sucatalog already exists and is up-to-date.
Dec 10 2018 16:57:18 -0600 DEBUG1: softwareupdate cmd: ['/usr/local/munki/ptyexec', '/usr/sbin/softwareupdate', '--verbose', '-i', u'macOS 10.14.2 Update- ']
Dec 10 2018 16:57:19 -0600 Installing macOS 10.14.2 Update
Dec 10 2018 16:57:48 -0600     Done.
Dec 10 2018 16:57:48 -0600     To install these updates, your computer must shut down. Your computer will automatically start up to finish installation.
Dec 10 2018 16:57:48 -0600     Installation will not complete successfully if you choose to restart your computer instead of shutting down.
Dec 10 2018 16:57:48 -0600     Please call halt(8) or select Shut Down from the Apple menu. To automate the shutdown process with softwareupdate(8), use --restart.
Dec 10 2018 16:57:48 -0600 DEBUG2: This update requires a shutdown...
Dec 10 2018 16:57:48 -0600 DEBUG2: InstallHistory.plist items:
Dec 10 2018 16:57:48 -0600 DEBUG1: Raw Apple Update install results: {'download': [], 'failures': [], 'installed': []}
Dec 10 2018 16:57:48 -0600 DEBUG2: Matched macOS 10.14.2 Update in InstallHistory.plist
Dec 10 2018 16:57:48 -0600 DEBUG2: Found shutdown flag...
Dec 10 2018 16:57:48 -0600 Finishing...
Dec 10 2018 16:57:48 -0600 Saving application inventory...
Dec 10 2018 16:57:48 -0600 DEBUG1: Getting info on currently installed applications...
Dec 10 2018 16:57:50 -0600     Performing postflight tasks...
Dec 10 2018 16:57:50 -0600     postflight stdout: 'cache' had errors during execution!
Dec 10 2018 16:57:50 -0600 ### Ending managedsoftwareupdate run ###
Dec 10 2018 16:57:50 -0600 Software installed or removed requires a restart.
Dec 10 2018 16:57:55 -0600 DEBUG1: Performing a shutdown...
Dec 10 2018 16:57:55 -0600     Restarting now.
Dec 10 2018 16:57:55 -0600 DEBUG1: Performing a regular restart...
```

The last line says `Performing a regular restart...` which isn't true since it's performing a shutdown. Perhaps a little bit of logic needs to be fixed so that line doesn't get written to the log?

Also instead of `Restarting now` it should say `Shutting down now.`

Good job @weswhet

**Comment by erikng** [comment]
Also, I think wiki should be updated for this with a note about T2 machines and that even though you send a halt, it still acts as a reboot and the T2 chip turns itself back on. I didn't realize that and thought my test was invalid and/or my device didn't actually need to use halt. It wasn't until I read the logs that I fully understood.

**Comment by gregneagle** [comment]
I've done some cleanup on this and made it available here: https://github.com/munki/munki/tree/softwareupdate-shutdown
@MagerValp might want to have a look.

**Comment by gregneagle** [comment]
A fun issue to consider: this PR was against Munki3dev -- I think we'll want to release this _before_ the Swift apps are ready to go, so we'll need to rebase it against master at some point.

**Comment by weswhet** [comment]
Do you want me to combine my changes and yours and file against master?

**Comment by gregneagle** [comment]
My changes are on top of yours, so they are already combined, no?

**Comment by weswhet** [comment]
Sure, either way should I open a new PR against master?

**Comment by gregneagle** [comment]
Closing this one in favor of #889

**Pull request closed without merging**